readME.md
 Welcome to block chain technology!!

Everything about Block Chain
[a link] https://blockgeeks.com/guides/what-is-blockchain-technology/

Ethereum:-
 [a link ] https://blockgeeks.com/guides/ethereum/
 
Smart Contracts:
  [a link] https://blockgeeks.com/guides/smart-contracts/

Solidity:

[a link] http://solidity.readthedocs.io/en/v0.4.20/


web3.js :

[a link] https://web3js.readthedocs.io/en/1.0/

nodeJS :
[a link] https://www.w3schools.com/nodejs/default.asp

Follow the steps:
step 1 :
    download gitbash for easy installation of packages required
    mkdir Remix_tool //create a directory called Remix_tool
    cd Remix_tool //change directory to Remix_tool
    git clone https://github.com/ethereum/browser-solidity //downloading solidity browser
    Install npm:
        sudo apt-get install nodejs //installing node.js
        sudo apt-get install npm // installing Node Package Manager
        sudo ln -s /usr/bin/nodejs /usr/bin/node //Create a symbolic link for node
    cd browser-solidity
        npm install
        npm run prepublish


step 2:
    clone the project and open the folder in gitbash
    now u need to install the packages , for that "npm install yarn" & "yarn init" commands are used
    install nodemon local server ("npm install nodemon")


step 3:
    run the solidity code in remix IDE and copy the ABI & contract Address and paste then in app.js 
    now in the console run the server (nodemon app.js)
    open the browser and follow the link (http://127.0.0.1:3000/account)


Private Block Chain:


1. Install geth cli 
    windows (download the .exe file from official site and install)
    ubuntu :
    Follow the commands:
        1. sudo apt-get install software-properties-common
        2. sudo add-apt-repository -y ppa:ethereum/ethereum
        3. sudo apt-get update
        4. sudo apt-get install ethereum


2. Open your gitbash and init the genesis file :
    geth --datadir "chain1" init "genesis.json"

3. Run the geth console in "8545" rpc port,
    geth --datadir "chain1" --port 30303 --networkid 1234 --rpc --rpcport "8545" --rpccorsdomain "*" console 2>console.log
    create few accounts
        personal.newAccount()
            //parsephrase (password)
        eth.setEtherbase(personal.listAccounts[0])
    start mining
        miner.start(4) //parameter indicates no. of threads for mining

3. Inthe remix IDE change the environment to web3 provider frrom javaScript





chain1/history
miner.start(1)
eth.nodeNumber
eth.blockNumber
miner.stop()
miner.start()[D1)
miner.start(1)
[A[B[1;5A[B[B[1eth.blockNumber
eth.blockNumber
miner.stop()
miner.start(1)
eth.blockNumber
miner.stop()
eth.blockNumber
the.accounts
eth.accounts
eth.blockNumber
loadScript()[D""dScript("SupplyChain.js")
loadScript("SupplyChain.js")
miner.start(1)
SupplyChain
SupplyChain.js
miner.stop()
personal.listAccounts
dsaf
miner.setEtherbase(eth.account[0])
[A[Bminer.setEtherbase(eth.accounts[0])
miner.setEtherBase(eth.accounts[0])
miner.setEtherbase(eth.accounts[2])[D[D[C0])
miner.setEtherbase(personal.listAccounts[0])
miner.start(1)
eth.blockNumber
miner.stop()
miner.start(1)
[A[Beth.blockNumber
miner.stop()
miner.start(1)
eth.blockNumber
eth.pendingTransactions
eth.blockNumber
miner.start(4)
eth.blockNumber
miner.stop()

webdeploy.js
var instructorContract = web3.eth.contract([{"constant":false,"inputs":[{"name":"name","type":"string"},{"name":"age","type":"uint256"}],"name":"setInstructor","outputs":[],"payable":false,"stateMutability":"nonpayable","type":"function"},{"constant":true,"inputs":[{"name":"","type":"uint256"}],"name":"instructors","outputs":[{"name":"_name","type":"string"},{"name":"_age","type":"uint256"}],"payable":false,"stateMutability":"view","type":"function"},{"constant":true,"inputs":[],"name":"id","outputs":[{"name":"","type":"uint256"}],"payable":false,"stateMutability":"view","type":"function"},{"constant":false,"inputs":[{"name":"id","type":"uint256"}],"name":"getInstructor","outputs":[{"name":"","type":"string"},{"name":"","type":"uint256"}],"payable":false,"stateMutability":"nonpayable","type":"function"}]);
var instructor = instructorContract.new(
   {
     from: web3.eth.accounts[0], 
     data: '0x606060405260008055341561001357600080fd5b6104a2806100226000396000f300606060405260043610610062576000357c0100000000000000000000000000000000000000000000000000000000900463ffffffff16806322faf03a146100675780638d532554146100cd578063af640d0f1461018d578063e77f9046146101b6575b600080fd5b341561007257600080fd5b6100cb600480803590602001908201803590602001908080601f01602080910402602001604051908101604052809392919081815260200183838082843782019150505050505091908035906020019091905050610259565b005b34156100d857600080fd5b6100ee60048080359060200190919050506102b8565b604051808060200183815260200182810382528481815460018160011615610100020316600290048152602001915080546001816001161561010002031660029004801561017d5780601f106101525761010080835404028352916020019161017d565b820191906000526020600020905b81548152906001019060200180831161016057829003601f168201915b5050935050505060405180910390f35b341561019857600080fd5b6101a06102db565b6040518082815260200191505060405180910390f35b34156101c157600080fd5b6101d760048080359060200190919050506102e1565b6040518080602001838152602001828103825284818151815260200191508051906020019080838360005b8381101561021d578082015181840152602081019050610202565b50505050905090810190601f16801561024a5780820380516001836020036101000a031916815260200191505b50935050505060405180910390f35b6000806000815480929190600101919050559050826001600083815260200190815260200160002060000190805190602001906102979291906103bd565b50816001600083815260200190815260200160002060010181905550505050565b600160205280600052604060002060009150905080600001908060010154905082565b60005481565b6102e961043d565b6000600160008481526020019081526020016000206000016001600085815260200190815260200160002060010154818054600181600116156101000203166002900480601f0160208091040260200160405190810160405280929190818152602001828054600181600116156101000203166002900480156103ad5780601f10610382576101008083540402835291602001916103ad565b820191906000526020600020905b81548152906001019060200180831161039057829003601f168201915b5050505050915091509150915091565b828054600181600116156101000203166002900490600052602060002090601f016020900481019282601f106103fe57805160ff191683800117855561042c565b8280016001018555821561042c579182015b8281111561042b578251825591602001919060010190610410565b5b5090506104399190610451565b5090565b602060405190810160405280600081525090565b61047391905b8082111561046f576000816000905550600101610457565b5090565b905600a165627a7a72305820c63692e41508d1bcdc1555d7da3e818f093423c9b69d7c710e826d871926c5870029', 
     gas: '4700000'
   }, function (e, contract){
    console.log(e, contract);
    if (typeof contract.address !== 'undefined') {
         console.log('Contract mined! address: ' + contract.address + ' transactionHash: ' + contract.transactionHash);
    }
 })

public/javaScript/script1.js
(function($) {
    "use strict"; // Start of use strict
  
    // Configure tooltips for collapsed side navigation
    $('.navbar-sidenav [data-toggle="tooltip"]').tooltip({
      template: '<div class="tooltip navbar-sidenav-tooltip" role="tooltip"><div class="arrow"></div><div class="tooltip-inner"></div></div>'
    })
  
    // Toggle the side navigation
    $("#sidenavToggler").click(function(e) {
      e.preventDefault();
      $("body").toggleClass("sidenav-toggled");
      $(".navbar-sidenav .nav-link-collapse").addClass("collapsed");
      $(".navbar-sidenav .sidenav-second-level, .navbar-sidenav .sidenav-third-level").removeClass("show");
    });
  
    // Force the toggled class to be removed when a collapsible nav link is clicked
    $(".navbar-sidenav .nav-link-collapse").click(function(e) {
      e.preventDefault();
      $("body").removeClass("sidenav-toggled");
    });
  
    // Prevent the content wrapper from scrolling when the fixed side navigation hovered over
    $('body.fixed-nav .navbar-sidenav, body.fixed-nav .sidenav-toggler, body.fixed-nav .navbar-collapse').on('mousewheel DOMMouseScroll', function(e) {
      var e0 = e.originalEvent,
        delta = e0.wheelDelta || -e0.detail;
      this.scrollTop += (delta < 0 ? 1 : -1) * 30;
      e.preventDefault();
    });
  
    // Scroll to top button appear
    $(document).scroll(function() {
      var scrollDistance = $(this).scrollTop();
      if (scrollDistance > 100) {
        $('.scroll-to-top').fadeIn();
      } else {
        $('.scroll-to-top').fadeOut();
      }
    });
  
    // Configure tooltips globally
    $('[data-toggle="tooltip"]').tooltip()
  
    // Smooth scrolling using jQuery easing
    $(document).on('click', 'a.scroll-to-top', function(event) {
      var $anchor = $(this);
      $('html, body').stop().animate({
        scrollTop: ($($anchor.attr('href')).offset().top)
      }, 1000, 'easeInOutExpo');
      event.preventDefault();
    });
  
    // Call the dataTables jQuery plugin
    $(document).ready(function() {
      $('#dataTable').DataTable();
    });
  
  })(jQuery); // End of use strict
  
  // Chart.js scripts
  // -- Set new default font family and font color to mimic Bootstrap's default styling
  Chart.defaults.global.defaultFontFamily = '-apple-system,system-ui,BlinkMacSystemFont,"Segoe UI",Roboto,"Helvetica Neue",Arial,sans-serif';
  Chart.defaults.global.defaultFontColor = '#292b2c';
  
  // -- Area Chart Example
  var ctx = document.getElementById("myAreaChart");
  var myLineChart = new Chart(ctx, {
    type: 'line',
    data: {
      labels: ["Mar 1", "Mar 2", "Mar 3", "Mar 4", "Mar 5", "Mar 6", "Mar 7", "Mar 8", "Mar 9", "Mar 10", "Mar 11", "Mar 12", "Mar 13"],
      datasets: [{
        label: "Sessions",
        lineTension: 0.3,
        backgroundColor: "rgba(2,117,216,0.2)",
        borderColor: "rgba(2,117,216,1)",
        pointRadius: 5,
        pointBackgroundColor: "rgba(2,117,216,1)",
        pointBorderColor: "rgba(255,255,255,0.8)",
        pointHoverRadius: 5,
        pointHoverBackgroundColor: "rgba(2,117,216,1)",
        pointHitRadius: 20,
        pointBorderWidth: 2,
        data: [10000, 30162, 26263, 18394, 18287, 28682, 31274, 33259, 25849, 24159, 32651, 31984, 38451],
      }],
    },
    options: {
      scales: {
        xAxes: [{
          time: {
            unit: 'date'
          },
          gridLines: {
            display: false
          },
          ticks: {
            maxTicksLimit: 7
          }
        }],
        yAxes: [{
          ticks: {
            min: 0,
            max: 40000,
            maxTicksLimit: 5
          },
          gridLines: {
            color: "rgba(0, 0, 0, .125)",
          }
        }],
      },
      legend: {
        display: false
      }
    }
  });
  
  // -- Bar Chart Example
  var ctx = document.getElementById("myBarChart");
  var myLineChart = new Chart(ctx, {
    type: 'bar',
    data: {
      labels: ["January", "February", "March", "April", "May", "June"],
      datasets: [{
        label: "Revenue",
        backgroundColor: "rgba(2,117,216,1)",
        borderColor: "rgba(2,117,216,1)",
        data: [4215, 5312, 6251, 7841, 9821, 14984],
      }],
    },
    options: {
      scales: {
        xAxes: [{
          time: {
            unit: 'month'
          },
          gridLines: {
            display: false
          },
          ticks: {
            maxTicksLimit: 6
          }
        }],
        yAxes: [{
          ticks: {
            min: 0,
            max: 15000,
            maxTicksLimit: 5
          },
          gridLines: {
            display: true
          }
        }],
      },
      legend: {
        display: false
      }
    }
  });
  
  // -- Pie Chart Example
  var ctx = document.getElementById("myPieChart");
  var myPieChart = new Chart(ctx, {
    type: 'pie',
    data: {
      labels: ["Blue", "Red", "Yellow", "Green"],
      datasets: [{
        data: [12.21, 15.58, 11.25, 8.32],
        backgroundColor: ['#007bff', '#dc3545', '#ffc107', '#28a745'],
      }],
    },
  });
  

app.js
const bodyParser = require('body-parser'); //Body parser is a middleware to handle POST data in Express.
const urlencodedParser = bodyParser.urlencoded({extended: true});
const express = require('express')
const sessions = require('express-session')
const app = express(); // routing
const cors = require('cors');
const web3 = require('web3');
const path = require('path');	
const http = require('http');
const Web3 = new web3();

Web3.setProvider(new Web3.providers.HttpProvider('http://localhost:8545'));
//Web3.eth.getAccounts()
var abi =[
	{
		"constant": true,
		"inputs": [],
		"name": "_p_id",
		"outputs": [
			{
				"name": "",
				"type": "uint256"
			}
		],
		"payable": false,
		"stateMutability": "view",
		"type": "function"
	},
	{
		"constant": true,
		"inputs": [
			{
				"name": "",
				"type": "uint256"
			}
		],
		"name": "tracks",
		"outputs": [
			{
				"name": "_product_id",
				"type": "uint256"
			},
			{
				"name": "_owner_id",
				"type": "uint256"
			},
			{
				"name": "_product_owner",
				"type": "address"
			},
			{
				"name": "_timeStamp",
				"type": "uint256"
			}
		],
		"payable": false,
		"stateMutability": "view",
		"type": "function"
	},
	{
		"constant": true,
		"inputs": [
			{
				"name": "",
				"type": "uint256"
			}
		],
		"name": "products",
		"outputs": [
			{
				"name": "_product_name",
				"type": "string"
			},
			{
				"name": "_product_cost",
				"type": "uint256"
			},
			{
				"name": "_product_specs",
				"type": "string"
			},
			{
				"name": "_product_review",
				"type": "string"
			},
			{
				"name": "_product_owner",
				"type": "address"
			},
			{
				"name": "_manufacture_date",
				"type": "uint256"
			}
		],
		"payable": false,
		"stateMutability": "view",
		"type": "function"
	},
	{
		"constant": true,
		"inputs": [
			{
				"name": "",
				"type": "uint256"
			}
		],
		"name": "participants",
		"outputs": [
			{
				"name": "_userName",
				"type": "string"
			},
			{
				"name": "_passWord",
				"type": "string"
			},
			{
				"name": "_address",
				"type": "address"
			},
			{
				"name": "_userType",
				"type": "string"
			}
		],
		"payable": false,
		"stateMutability": "view",
		"type": "function"
	},
	{
		"constant": true,
		"inputs": [],
		"name": "_t_id",
		"outputs": [
			{
				"name": "",
				"type": "uint256"
			}
		],
		"payable": false,
		"stateMutability": "view",
		"type": "function"
	},
	{
		"constant": true,
		"inputs": [],
		"name": "_u_id",
		"outputs": [
			{
				"name": "",
				"type": "uint256"
			}
		],
		"payable": false,
		"stateMutability": "view",
		"type": "function"
	},
	{
		"constant": false,
		"inputs": [
			{
				"name": "trck_id",
				"type": "uint256"
			}
		],
		"name": "getProduct_trackindex",
		"outputs": [
			{
				"name": "",
				"type": "uint256"
			},
			{
				"name": "",
				"type": "uint256"
			},
			{
				"name": "",
				"type": "address"
			},
			{
				"name": "",
				"type": "uint256"
			}
		],
		"payable": false,
		"stateMutability": "nonpayable",
		"type": "function"
	},
	{
		"constant": false,
		"inputs": [
			{
				"name": "uid",
				"type": "uint256"
			},
			{
				"name": "uname",
				"type": "string"
			},
			{
				"name": "pass",
				"type": "string"
			},
			{
				"name": "utype",
				"type": "string"
			}
		],
		"name": "userLogin",
		"outputs": [
			{
				"name": "",
				"type": "bool"
			}
		],
		"payable": false,
		"stateMutability": "nonpayable",
		"type": "function"
	},
	{
		"constant": false,
		"inputs": [
			{
				"name": "own_id",
				"type": "uint256"
			},
			{
				"name": "name",
				"type": "string"
			},
			{
				"name": "p_cost",
				"type": "uint256"
			},
			{
				"name": "p_specs",
				"type": "string"
			},
			{
				"name": "p_review",
				"type": "string"
			}
		],
		"name": "newProduct",
		"outputs": [
			{
				"name": "",
				"type": "uint256"
			}
		],
		"payable": false,
		"stateMutability": "nonpayable",
		"type": "function"
	},
	{
		"constant": false,
		"inputs": [
			{
				"name": "user1_id",
				"type": "uint256"
			},
			{
				"name": "user2_id",
				"type": "uint256"
			},
			{
				"name": "prod_id",
				"type": "uint256"
			}
		],
		"name": "transferOwnership_product",
		"outputs": [
			{
				"name": "",
				"type": "bool"
			}
		],
		"payable": false,
		"stateMutability": "nonpayable",
		"type": "function"
	},
	{
		"constant": false,
		"inputs": [
			{
				"name": "name",
				"type": "string"
			},
			{
				"name": "pass",
				"type": "string"
			},
			{
				"name": "u_add",
				"type": "address"
			},
			{
				"name": "utype",
				"type": "string"
			}
		],
		"name": "createParticipant",
		"outputs": [
			{
				"name": "",
				"type": "uint256"
			}
		],
		"payable": false,
		"stateMutability": "nonpayable",
		"type": "function"
	},
	{
		"constant": false,
		"inputs": [
			{
				"name": "p_id",
				"type": "uint256"
			}
		],
		"name": "getParticipant",
		"outputs": [
			{
				"name": "",
				"type": "string"
			},
			{
				"name": "",
				"type": "address"
			},
			{
				"name": "",
				"type": "string"
			}
		],
		"payable": false,
		"stateMutability": "nonpayable",
		"type": "function"
	},
	{
		"constant": false,
		"inputs": [
			{
				"name": "prod_id",
				"type": "uint256"
			}
		],
		"name": "getProduct_details",
		"outputs": [
			{
				"name": "",
				"type": "string"
			},
			{
				"name": "",
				"type": "uint256"
			},
			{
				"name": "",
				"type": "string"
			},
			{
				"name": "",
				"type": "string"
			},
			{
				"name": "",
				"type": "address"
			},
			{
				"name": "",
				"type": "uint256"
			}
		],
		"payable": false,
		"stateMutability": "nonpayable",
		"type": "function"
	}
];
var contadd = "0xffb7eebc6325b8da8671a3a1252cbb9c46a77e32";

var session;

app.use(sessions({
	secret:'$@%*^gjb&@^&nhv' ,
	resave: false ,
	saveUninitialized: true
}))

var con = Web3.eth.contract(abi).at(contadd);
app.use(cors());  //cross resourse sharing
app.use(bodyParser.urlencoded({ extended: true }));
app.use(bodyParser.json());
//console.log(app);
app.get('/newProduct',(req,res)=>{
	res.sendFile(__dirname + '/public/newProduct.html');
});
//
app.get('/transfer',(req,res)=>{
	res.sendFile(__dirname + '/public/transferProduct.html');
});
app.get('/account',(req, res)=> {
    res.sendFile(__dirname + '/public/register.html');
});
app.get('/index',(req, res)=> {
    res.sendFile(__dirname + '/public/index.html');
});
app.get('/login',(req, res)=> {
	session = req.session;
	if(session.uniqueID){
		res.redirect('/redirects');
	}
    res.sendFile(__dirname + '/public/login.html');
});
app.get('/logout',(req,res)=>{
	req.session.destroy();
	res.redirect('/login');
   });
   /*app.get('/redirect',(req,res)=>{
	   console.log
	   session = req.session;
	   console.log(session);
		if(session.uniqueID){
		   res.sendFile(__dirname + '/public/traceProduct.html');
		}else{
		   res.redirect('/login');
		}
   });*/


   app.get('/trace',(req,res)=>{
	   session = req.session;
	   if(!session.uniqueID){
		   res.redirect('/login');
	   }
	   
	   res.sendFile(__dirname + '/public/traceProduct.html');
	   console.log(session.uniqueID);
   });
   app.post('/transfer',(req,res)=>{
	   let response =[];
		let acc_id1 = parseInt(req.body.aid1);
		let address1= con.getParticipant.call(acc_id1);
		let add1 = address1[1];
		//response.push(add1);
		let acc_id2 = parseInt(req.body.aid2);
		let prd = parseInt(req.body.pid);
		let trns1 = con.transferOwnership_product.call(acc_id1,acc_id2,prd,{from:add1,gas:3000000});
		let trns = con.transferOwnership_product(acc_id1,acc_id2,prd,{from:add1,gas:3000000});
	//	if(trns){
		
		let track_id =parseInt(con._t_id.call());
		response.push(track_id);

		res.send(response);
		//res.send("transfered...");
		//}else{
		//	res.send("error...");
		//}
		//res.send("hey....");
});
   app.post('/trace', (req, res)=>{
		//let product_id = parseInt(req.body.pid);
		let response =[];
		let index = parseInt(req.body.tid);

		//let track_id =parseInt(con._t_id.call());
		let trackDetails = con.getProduct_trackindex.call(index);
		let track_id =parseInt(con._t_id.call());
		response.push(track_id);
		response.push(trackDetails[0]);
		response.push(trackDetails[1]);
		response.push(trackDetails[2]);
		response.push(trackDetails[3]);

		res.send(response);
   });
app.post('/login',(req,res)=>{
	session = req.session;
	if(session.uniqueID){
		res.redirect('/trace');
	}
	//try{
		var utype =  req.body.utype;
		var uid =parseInt(req.body.u_id);
		var name = req.body.name;
		var pass = req.body.pass;
		var p = con.userLogin.call(uid,name,pass,utype);
		//console.log("T:", types," :", stkid," ", password, "P: ", p);
		//session = req.session;
		if(p == true){
				session.uniqueID = req.body.name;	
				//res.redirect('/redirect');
				res.send("Login sucessfull");
		}else{
			//res.redirect('/redirect');
			res.send("Login Failed. Please try again.");
		}

	//}catch(error){
		//res.send("Some error occured.");
	//}
});

app.post('/newProduct',(req,res)=>{
	
	let userid = req.body.uid;
	let product_name = req.body.prdt_name;
	let prdt_cost = req.body.prdt_cost;
	let prdt_specs = req.body.prdt_specs;
	let prdt_desc = req.body.prdt_desc;

	let invoke = con.newProduct.call(userid,product_name,prdt_cost,prdt_specs,prdt_desc,{from:Web3.eth.accounts[0],gas:3000000});
	let invoke1 = con.newProduct(userid,product_name,prdt_cost,prdt_specs,prdt_desc,{from:Web3.eth.accounts[0],gas:3000000});
	if(parseInt(invoke1) != 0){
		let count =parseInt(con._p_id.call());
		let getProduct = con.getProduct_details.call(parseInt(count-1));
		let details=[]
	
		details.push(count-1);
		details.push(getProduct[0]);
		details.push(getProduct[1]);
		details.push(getProduct[2]);
		details.push(getProduct[3]);
		details.push(getProduct[4]);
		details.push(getProduct[5]);
	 
		res.send(details);
	}
	else{
		res.send("Invalid User");
	}
	
	
});
app.post('/account',(req ,res )=>{
	//if(!err){
		let response =[];
		let uname =req.body.name;
		//let x=1;
		let upass =req.body.pass;
		let utype = req.body.utype;
		let count =parseInt(con._u_id.call());
		let u_address =Web3.eth.accounts[count]; 
		let f_invoke = con.createParticipant.call(uname,upass,u_address,utype,{from:Web3.eth.accounts[0],gas:3000000});
		let f_invoke1 = con.createParticipant(uname,upass,u_address,utype,{from:Web3.eth.accounts[0],gas:3000000});
		let count1 =parseInt(con._u_id.call());
		let out =con.getParticipant.call(count1-1);
		//accnts.push(Web3.eth.accounts);
		response.push(f_invoke1);
		response.push(out);
		
		res.send(response);
		//res.redirect('http://127.0.0.1:3000/login');
	//}
    
});
app.listen(3000); 

